File: src/sniper_range/projectile.py
"""Visible tracer that integrates ballistics and raycasts to avoid tunneling."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ursina import Entity, Vec3, color, destroy, raycast, time

from sniper_range import config
from sniper_range.ballistics import Vec3 as PhysVec3
from sniper_range.ballistics import initial_velocity, step

if TYPE_CHECKING:
    from sniper_range.app import Game


def _to_phys(vector: Vec3) -> PhysVec3:
    return PhysVec3(float(vector.x), float(vector.y), float(vector.z))


def _to_ursina(vector: PhysVec3) -> Vec3:
    return Vec3(vector.x, vector.y, vector.z)


class Projectile(Entity):
    def __init__(
        self,
        game: Game,
        origin: Vec3,
        direction: Vec3,
        muzzle_velocity: float = config.MUZZLE_VELOCITY,
    ) -> None:
        super().__init__(
            model="cube",
            color=color.rgb(1.0, 0.72, 0.18),
            position=origin,
            scale=config.TRACER_SCALE,
            collider=None,
        )
        self.game = game
        self.phys_position = _to_phys(origin)
        self.phys_velocity = initial_velocity(_to_phys(direction), muzzle_velocity)
        self.age = 0.0
        self.look_at(self.position + _to_ursina(self.phys_velocity))

    def update(self) -> None:
        dt = min(time.dt, config.MAX_PHYSICS_DT)
        self.age += dt
        if self.age >= config.PROJECTILE_LIFETIME:
            destroy(self)
            return

        old_position = self.phys_position
        self.phys_position, self.phys_velocity = step(
            self.phys_position,
            self.phys_velocity,
            self.game.wind,
            dt,
            config.GRAVITY,
        )

        travel = self.phys_position - old_position
        distance = travel.length()
        if distance > 0.0:
            # Sweep a ray along the integrated segment so a 500 m/s tracer
            # cannot skip through a thin steel plate between two frames.
            hit = raycast(
                _to_ursina(old_position),
                _to_ursina(travel.normalized()),
                distance=distance,
                ignore=[*self.game.projectile_ignore, self],
                debug=False,
            )
            if hit.hit:
                self._impact(hit)
                return

        self.position = _to_ursina(self.phys_position)
        if self.phys_velocity.length() > 0.0:
            self.look_at(self.position + _to_ursina(self.phys_velocity))

    def _impact(self, hit) -> None:
        point = hit.world_point
        shot_distance = (point - self.game.player.world_position).length()
        target = hit.entity
        handler = getattr(target, "on_bullet_hit", None) if target is not None else None
        result = handler(point) if callable(handler) else None
        if result is None:
            _spawn_impact_mark(point, hit.world_normal)

        scored = bool(result is not None and result.scored)
        self.game.report_impact(
            hit=scored,
            distance=shot_distance,
            point=point,
            label=result.label if result is not None else "",
        )
        self.game.notify_impact(point)
        destroy(self)


def _spawn_impact_mark(point: Vec3, normal: Vec3) -> None:
    offset = (normal.normalized() * 0.04) if normal.length() else Vec3(0, 0.04, 0)
    mark = Entity(
        model="sphere",
        color=color.rgb(0.12, 0.1, 0.08),
        position=point + offset,
        scale=config.IMPACT_MARK_SCALE,
    )
    destroy(mark, delay=config.IMPACT_MARK_LIFETIME)


def spawn_projectile(
    game: Game,
    origin: Vec3,
    direction: Vec3,
    muzzle_velocity: float = config.MUZZLE_VELOCITY,
) -> Projectile:
    return Projectile(game, origin, direction, muzzle_velocity)

File: web/src/projectile.ts
/** Visible tracer that integrates ballistics and raycasts to avoid tunneling. */

import {
  BoxGeometry,
  Color,
  Mesh,
  MeshBasicMaterial,
  Raycaster,
  Vector3,
} from "three";

import { Vec3, initialVelocity, step } from "./ballistics";
import {
  GRAVITY,
  IMPACT_MARK_LIFETIME,
  IMPACT_MARK_SCALE,
  PROJECTILE_LIFETIME,
  TRACER_SCALE,
} from "./config";
import type { Game } from "./game";
import type { HitResult, SteelTarget } from "./targets";

const TRACER = new Color(1.0, 0.72, 0.18);
const MARK = new Color(0.12, 0.1, 0.08);

const _from = new Vector3();
const _dir = new Vector3();
const _look = new Vector3();
const raycaster = new Raycaster();

export class Projectile {
  readonly mesh: Mesh;
  private physPosition: Vec3;
  private physVelocity: Vec3;
  private age = 0;
  alive = true;

  constructor(
    private readonly game: Game,
    origin: Vec3,
    direction: Vec3,
    muzzleVelocity: number,
  ) {
    this.mesh = new Mesh(
      new BoxGeometry(TRACER_SCALE[0], TRACER_SCALE[1], TRACER_SCALE[2]),
      new MeshBasicMaterial({ color: TRACER }),
    );
    this.mesh.position.set(origin.x, origin.y, origin.z);
    this.physPosition = origin;
    this.physVelocity = initialVelocity(direction, muzzleVelocity);
    this.orient();
    game.scene.add(this.mesh);
    game.projectileIgnore.push(this.mesh);
  }

  update(dt: number): void {
    if (!this.alive) {
      return;
    }
    this.age += dt;
    if (this.age >= PROJECTILE_LIFETIME) {
      this.destroy();
      return;
    }

    const oldPosition = this.physPosition;
    [this.physPosition, this.physVelocity] = step(
      this.physPosition,
      this.physVelocity,
      this.game.wind,
      dt,
      GRAVITY,
    );
    const travel = this.physPosition.sub(oldPosition);
    const distance = travel.length();
    if (distance > 0) {
      _from.set(oldPosition.x, oldPosition.y, oldPosition.z);
      const n = travel.normalized();
      _dir.set(n.x, n.y, n.z);
      raycaster.far = distance;
      raycaster.set(_from, _dir);
      const hits = raycaster.intersectObjects(this.game.colliders, true);
      const hit = hits.find((entry) => !this.game.projectileIgnore.includes(entry.object));
      if (hit !== undefined) {
        this.impact(hit.point, hit.normal ?? new Vector3(0, 1, 0), hit.object);
        return;
      }
    }

    this.mesh.position.set(this.physPosition.x, this.physPosition.y, this.physPosition.z);
    this.orient();
  }

  private orient(): void {
    if (this.physVelocity.length() <= 0) {
      return;
    }
    _look.set(
      this.mesh.position.x + this.physVelocity.x,
      this.mesh.position.y + this.physVelocity.y,
      this.mesh.position.z + this.physVelocity.z,
    );
    this.mesh.lookAt(_look);
  }

  private impact(point: Vector3, normal: Vector3, object: { userData: { target?: SteelTarget } }): void {
    const playerPos = this.game.player.worldPosition;
    const shotDistance = point.distanceTo(playerPos);
    const target = findTarget(object);
    let result: HitResult | null = null;
    if (target !== undefined) {
      result = target.onBulletHit();
    }
    if (result === null) {
      spawnImpactMark(this.game, point, normal);
    }
    this.game.reportImpact({
      hit: result !== null && result.scored,
      distance: shotDistance,
      label: result?.label ?? "",
    });
    this.destroy();
  }

  private destroy(): void {
    if (!this.alive) {
      return;
    }
    this.alive = false;
    this.game.scene.remove(this.mesh);
    this.mesh.geometry.dispose();
    const ignore = this.game.projectileIgnore;
    const index = ignore.indexOf(this.mesh);
    if (index >= 0) {
      ignore.splice(index, 1);
    }
  }
}

function findTarget(object: { userData: { target?: SteelTarget }; parent?: unknown }): SteelTarget | undefined {
  let current: { userData: { target?: SteelTarget }; parent?: unknown } | null = object;
  while (current !== null) {
    if (current.userData.target !== undefined) {
      return current.userData.target;
    }
    current = (current.parent as typeof current) ?? null;
  }
  return undefined;
}

function spawnImpactMark(game: Game, point: Vector3, normal: Vector3): void {
  const mark = new Mesh(
    new BoxGeometry(IMPACT_MARK_SCALE, IMPACT_MARK_SCALE, IMPACT_MARK_SCALE),
    new MeshBasicMaterial({ color: MARK }),
  );
  const offset = normal.clone().normalize().multiplyScalar(0.04);
  mark.position.copy(point).add(offset);
  game.scene.add(mark);
  window.setTimeout(() => {
    game.scene.remove(mark);
    mark.geometry.dispose();
  }, IMPACT_MARK_LIFETIME * 1000);
}

export function spawnProjectile(
  game: Game,
  origin: Vec3,
  direction: Vec3,
  muzzleVelocity: number,
): Projectile {
  const projectile = new Projectile(game, origin, direction, muzzleVelocity);
  game.projectiles.push(projectile);
  return projectile;
}
